File: golang-devildyw/middle/L49字母异位词分组.go
package main

import "sort"

func main() {

}

func groupAnagrams(strs []string) [][]string {
	if len(strs) == 1 {
		return [][]string{strs}
	}
	result := make([][]string, 0)
	// 对每个单词排序 这个挺耗时的吧 感觉
	m := make(map[string][]string)
	for _, str := range strs {
		// 对 str 排序
		bytes := []byte(str)
		sort.Slice(bytes, func(i, j int) bool {
			return bytes[i] < bytes[j]
		})

		newStr := string(bytes)

		if value, exist := m[newStr]; exist {
			value = append(value, str)
			m[newStr] = value
		} else {
			strList := []string{str}
			m[newStr] = strList
		}
	}

	for _, value := range m {
		result = append(result, value)
	}
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/medium/L49字母异位词分组.java
package top.dyw.leetcode.medium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class L49字母异位词分组 {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList();

        HashMap<String, List<String>> map = new HashMap();
        for (String str : strs) {
            char[] bytes = str.toCharArray();
            Arrays.sort(bytes);
            String newStr = new String(bytes);

            List<String> list = map.getOrDefault(newStr, new ArrayList<String>());
            list.add(str);
            map.put(newStr, list);
        }

        for (String key : map.keySet()) {
            result.add(map.get(key));
        }

        return result;
    }

}

